### PR TITLE
feature: better support for cockroachdb

### LINKED ISSUE #7573: CockroachDB Geography Type
## Feature Description

TypeORM should support the GEOGRAPHY and GEOMETRY type as it does for Postgres. (it is supported by cockroachDB: https://www.cockroachlabs.com/docs/v20.2/spatial-data.html)

### The Problem

TypeORM doesn't support the GEOGRAPHY type for CockroachDB

### The Solution

Add the type to the supported types

### Relevant Database Driver(s)

- [ ] `aurora-data-api`
- [ ] `aurora-data-api-pg`
- [ ] `better-sqlite3`
- [X] `cockroachdb`
- [ ] `cordova`
- [ ] `expo`
- [ ] `mongodb`
- [ ] `mysql`
- [ ] `nativescript`
- [ ] `oracle`
- [ ] `postgres`
- [ ] `react-native`
- [ ] `sap`
- [ ] `sqlite`
- [ ] `sqlite-abstract`
- [ ] `sqljs`
- [ ] `sqlserver`

### Are you willing to resolve this issue by submitting a Pull Request?

- [ ] Yes, I have the time, and I know how to start.
- [X] Yes, I have the time, but I don't know how to start. I would need guidance.
- [ ] No, I don't have the time, although I believe I could do it if I had the time...
- [ ] No, I don't have the time and I wouldn't even know how to start.

Thank you!

### LINKED ISSUE #8039: Add support for ENUMs for cockroachdb
## Feature Description

CockroachDB supports enums very similarly to how Postgres does. The TypeORM driver for CockroachDB should recognize that.

### The Problem

Currently, TypeORM does not recognize enum as a supported type for CockroachDB, so this feature is not available to users.

### The Solution

Support enums for CockroachDB just as they are for Postgres. See https://github.com/typeorm/typeorm/commit/43e52c57903fa9c86c947f3c8fc45c718ee4f2e1 for initial PG enum support as well as the current code.

### Considered Alternatives

Use string constants,

### Additional Context

This was requested on the cockroachdb repo here: https://github.com/cockroachdb/cockroach/issues/68368

### Relevant Database Driver(s)

| DB Type              | Relevant      |
| ---                  | ---           |
| `aurora-data-api`    | no            |
| `aurora-data-api-pg` | no            |
| `better-sqlite3`     | no            |
| `cockroachdb`        | yes            |
| `cordova`            | no            |
| `expo`               | no            |
| `mongodb`            | no            |
| `mysql`              | no            |
| `nativescript`       | no            |
| `oracle`             | no            |
| `postgres`           | no            |
| `react-native`       | no            |
| `sap`                | no            |
| `sqlite`             | no            |
| `sqlite-abstract`    | no            |
| `sqljs`              | no            |
| `sqlserver`          | no            |

### Are you willing to resolve this issue by submitting a Pull Request?

 - ✖️ Yes, I have the time, and I know how to start.
 - ✅ Yes, I have the time, but I don't know how to start. I would need guidance.
 - ✖️ No, I don’t have the time, but I can support (using donations) development.
 - ✖️ No, I don’t have the time and I’m okay to wait for the community / maintainers to resolve this issue.

### LINKED ISSUE #9068: Data type "enum" in "User.gender" is not supported by "cockroachdb"
## Issue Description

It seems that enum types does not work with cochroachdb as driver. If you use postgres it works.

### Expected Behavior

It should support / create the custom enum type just like it does in postgres. [Enums are supported by cockroachdb](https://www.cockroachlabs.com/docs/stable/enum.html).

If you set the driver to `postgres` against an cockroacbdb instance, it will work just fine.

### Actual Behavior

This error is outputted:

```
[Nest] 64905  - 06/02/2022, 2:14:26 PM   ERROR [ExceptionHandler] Data type "enum" in "User.gender" is not supported by "cockroachdb" database.
DataTypeNotSupportedError: Data type "enum" in "User.gender" is not supported by "cockroachdb" database.
    at DataTypeNotSupportedError.TypeORMError [as constructor] (/Users/mf/Projects/project/backend/src/error/TypeORMError.ts:7:9)
    at new DataTypeNotSupportedError (/Users/mf/Projects/project/backend/src/error/DataTypeNotSupportedError.ts:8:9)
    at /Users/mf/Projects/project/backend/src/metadata-builder/EntityMetadataValidator.ts:89:27
    at Array.forEach (<anonymous>)
    at EntityMetadataValidator.validate (/Users/mf/Projects/project/backend/src/metadata-builder/EntityMetadataValidator.ts:86:36)
    at /Users/mf/Projects/project/backend/src/metadata-builder/EntityMetadataValidator.ts:46:56
    at Array.forEach (<anonymous>)
    at EntityMetadataValidator.validateMany (/Users/mf/Projects/project/backend/src/metadata-builder/EntityMetadataValidator.ts:46:25)
    at Connection.<anonymous> (/Users/mf/Projects/project/backend/src/connection/Connection.ts:533:33)
    at step (/Users/mf/Projects/project/backend/node_modules/tslib/tslib.js:143:27)
```

If you first set the driver to be `postgres`, and then go into your cockroachdb, and call `SHOW ENUMS;`, you'll actually see the enums:

| schema | name                                    | values                  | owner |
|--------|-----------------------------------------|-------------------------|-------|
| public | user_gender_enum              | {Male,Female,NonBinary} | root  |

UPDATE: Not sure exactly when it was created, however, now I cannot recreate it. Tried deleting the database, and starting the app again, this time the same error occurs, but `SHOW ENUMS;` reveals nothing

### Steps to Reproduce

I have defined the enum like so:
```typescript
export enum Gender {
  Male = 'Male',
  Female = 'Female',
  NonBinary = 'NonBinary',
}

@Column({
  type: 'enum',
  enum: Gender,
  nullable: true,
})
gender: Gender;
```

### My Environment

| Dependency          | Version  |
| ---                 | ---      |
| Operating System    |          |
| Node.js version     | 16.14.2  |
| Typescript version  | 4.6.3  |
| TypeORM version     | 0.2.45  | 
| CockroachDB     | v21.2  | 

### Relevant Database Driver(s)

| DB Type           | Reproducible  |
|-------------------| ---           |
| `aurora-mysql`    | no            |
| `aurora-postgres` | no            |
| `better-sqlite3`  | no            |
| `cockroachdb`     | yes            |
| `cordova`         | no            |
| `expo`            | no            |
| `mongodb`         | no            |
| `mysql`           | no            |
| `nativescript`    | no            |
| `oracle`          | no            |
| `postgres`        | no            |
| `react-native`    | no            |
| `sap`             | no            |
| `spanner`         | no            |
| `sqlite`          | no            |
| `sqlite-abstract` | no            |
| `sqljs`           | no            |
| `sqlserver`       | no            |

### Are you willing to resolve this issue by submitting a Pull Request?

 - ✖️ Yes, I have the time, and I know how to start.
 - ✖️ Yes, I have the time, but I don't know how to start. I would need guidance.
 - ✅ No, I don’t have the time, but I can support (using donations) development.
 - ✖️ No, I don’t have the time and I’m okay to wait for the community / maintainers to resolve this issue.

### LINKED ISSUE #9127: cockrachdb: DROP TYPE CASCADE is not yet supported
## Issue Description

When using cockroachdb with enums the app won't work - despite syncronize is disabled. It tries to create an enum type, and uses DROP CASCADE, which is not supported by cockroachdb.

### Expected Behavior

Ability to disable `DROP TYPE IF EXISTS .... CASCADE` - I'd like to write the migrations/database schema myself entirely.

If I could use the cockroachdb driver, it would be better: https://github.com/typeorm/typeorm/issues/9068

### Actual Behavior

```console
query failed: DROP TYPE IF EXISTS "public"."gender_enum" CASCADE;
error: error: unimplemented: DROP TYPE CASCADE is not yet supported
    at Parser.parseErrorMessage (/Users/me/Projects/project/backend/node_modules/pg-protocol/src/parser.ts:369:69)
    at Parser.handlePacket (/Users/me/Projects/project/backend/node_modules/pg-protocol/src/parser.ts:188:21)
    at Parser.parse (/Users/me/Projects/project/backend/node_modules/pg-protocol/src/parser.ts:103:30)
    at Socket.<anonymous> (/Users/me/Projects/project/backend/node_modules/pg-protocol/src/index.ts:7:48)
    at Socket.emit (node:events:526:28)
    at Socket.emit (node:domain:475:12)
    at addChunk (node:internal/streams/readable:315:12)
    at readableAddChunk (node:internal/streams/readable:289:9)
    at Socket.Readable.push (node:internal/streams/readable:228:10)
    at TCP.onStreamRead (node:internal/stream_base_commons:190:23) {
  length: 213,
  severity: 'ERROR',
  code: '0A000',
  detail: undefined,
  hint: 'You have attempted to use a feature that is not yet implemented.\n' +
    'See: https://go.crdb.dev/issue-v/51480/v22.1',
  position: undefined,
  internalPosition: undefined,
  internalQuery: undefined,
  where: undefined,
  schema: undefined,
  table: undefined,
  column: undefined,
  dataType: undefined,
  constraint: undefined,
  file: 'drop_type.go',
  line: '53',
  routine: 'DropType'
}
```

### Steps to Reproduce

1. Create an entity
2. Assign an custom enum type to the entity
3. Set the driver to `postgres` (otherwise will it simply complain with: https://github.com/typeorm/typeorm/issues/9068)

```typescript
export enum Gender {
  Male = 'Male',
  Female = 'Female',
  NonBinary = 'NonBinary',
}

@Column({
  type: 'enum',
  enum: Gender,
  nullable: true,
})
@Index()
gender: Gender;
```

### My Environment

| Dependency          | Version  |
| ---                 | ---      |
| Operating System    |          |
| Node.js version     | v16.14.2  | 
| Typescript version  | 4.6.3  | 
| TypeORM version     | 0.3.6  | 

### Relevant Database Driver(s)

| DB Type           | Reproducible  |
|-------------------| ---           |
| `aurora-mysql`    | no            |
| `aurora-postgres` | no            |
| `better-sqlite3`  | no            |
| `cockroachdb`     | no            |
| `cordova`         | no            |
| `expo`            | no            |
| `mongodb`         | no            |
| `mysql`           | no            |
| `nativescript`    | no            |
| `oracle`          | no            |
| `postgres`        | yes            |
| `react-native`    | no            |
| `sap`             | no            |
| `spanner`         | no            |
| `sqlite`          | no            |
| `sqlite-abstract` | no            |
| `sqljs`           | no            |
| `sqlserver`       | no            |

### Are you willing to resolve this issue by submitting a Pull Request?

 - ✖️ Yes, I have the time, and I know how to start.
 - ✅ Yes, I have the time, but I don't know how to start. I would need guidance.
 - ✅ No, I don’t have the time, but I can support (using donations) development.
 - ✖️ No, I don’t have the time and I’m okay to wait for the community / maintainers to resolve this issue.

### LINKED ISSUE #9199: Add support for CockroachDB 'upsert'
## Feature Description
Should add support for CockroachDB's [UPSERT](https://www.cockroachlabs.com/docs/v22.1/upsert#upsert-vs-insert-on-conflict) query which has better performance, and works better with Composite Primary keys.

### The Problem

Cockroach has a dedicated `UPSERT` method. Currently TypeORM's `.upsert()` method calls an `INSERT ON CONFLICT` query instead, which has worse performance and causes some bugs in our case.

### The Solution

- Add a `upsert` option to `UpsertType` (`src/driver/types/UpsertType.ts`).
- Rename `supportedUpsertType` to `supportedUpsertTypes` and make it an array (since technically cockroachdb supports both approaches)
- Add an optional `UpsertType` to `InsertOrUpdateOptions` that let's the user define a custom Upsert for a given `.upsert()` query
- The first `UpsertType` value in the driver will be the default `UpsertType` if none was provided in the query. For backwards compatibility, to have `.upsert()` work the same as it does currently.
- Tweak `InsertQueryBuilder.ts` to output `UPSERT` instead of `INSERT` if `UpsertType` is equal to `upsert`. Also omit the `ON CONFLICT` part that is not supported in an `UPSERT` query

### Considered Alternatives

Writing an raw query, but that kinda defeats the purpose

### Relevant Database Driver(s)

| DB Type           | Relevant      |
|-------------------| ---           |
| `aurora-mysql`    | no            |
| `aurora-postgres` | no            |
| `better-sqlite3`  | no            |
| `cockroachdb`     | yes            |
| `cordova`         | no            |
| `expo`            | no            |
| `mongodb`         | no            |
| `mysql`           | no            |
| `nativescript`    | no            |
| `oracle`          | no            |
| `postgres`        | no            |
| `react-native`    | no            |
| `sap`             | no            |
| `spanner`         | no            |
| `sqlite`          | no            |
| `sqlite-abstract` | no            |
| `sqljs`           | no            |
| `sqlserver`       | no            |

### Are you willing to resolve this issue by submitting a Pull Request?

 - ✅ Yes, I have the time, and I know how to start.
 - ✖️ Yes, I have the time, but I don't know how to start. I would need guidance.
 - ✖️ No, I don’t have the time, but I can support (using donations) development.
 - ✖️ No, I don’t have the time and I’m okay to wait for the community / maintainers to resolve this issue.

### PR DESCRIPTION
**Pull request by simplenotezy**
### Description of change

[cockroachdb supports enums](https://www.cockroachlabs.com/docs/stable/data-types.html#supported-types) - but it is not listed as an supported type in `src/driver/cockroachdb/CockroachDriver.ts`

### Features
- [x] Support for `enum` types (fixes #8039, fixes #9068)
- [x] Support for `geometry` / `geography` types (fixes #7573)
- [x] Support for `UPSERT` (fixes #9199)
- [x] Support for [Time Travelling Queries](https://www.cockroachlabs.com/docs/stable/as-of-system-time.html) 

### Bugfixes
- [x] [prevent error if trying to close already closed connection](https://github.com/typeorm/typeorm/pull/9128/commits/6993d4850cb00c57398055f04119c96af54547f6)
- [x] [bugfix: type casts to string when using ANY](https://github.com/typeorm/typeorm/pull/9128/commits/d5604e525b87efabf9dd89b00c9816cf3b9f2722)
- [x] fixes #9127 

### Changes
- [x] [feature: added cockrachodb as part of postgres family](https://github.com/typeorm/typeorm/pull/9128/commits/f703edb4bfc7fbaa21a2f6d78f97d16f05411b1e)

### Pull-Request Checklist

- [x] Code is up-to-date with the `master` branch
- [x] `npm run format` to apply prettier formatting
- [x] `npm run test` passes with this change
- [x] This pull request links relevant issues as `Fixes #0000`
- [ ] There are new or updated unit tests validating the change
- [x] Documentation has been updated to reflect this change
- [x] The new commits follow conventions explained in [CONTRIBUTING.md](https://github.com/typeorm/typeorm/blob/master/CONTRIBUTING.md)

### CODE REVIEW

File: src/query-builder/QueryBuilder.ts [hunk 1]
```diff
@@ -1042,6 +1042,10 @@ export abstract class QueryBuilder<Entity> {
                     .slice(1)
                     .join(", ")})`
             case "any":
+                if (driver.options.type === "cockroachdb") {
+                    return `${condition.parameters[0]}::STRING = ANY(${condition.parameters[1]}::STRING[])`
```
Comment (pleerock) [inline src/query-builder/QueryBuilder.ts:1046]: why do you cast it to string? what if parameter is not a string?

File: src/driver/cockroachdb/CockroachQueryRunner.ts [hunk 2]
```diff
@@ -122,6 +122,10 @@ export class CockroachQueryRunner
      * You cannot use query runner methods once its released.
      */
     release(): Promise<void> {
+        if (this.isReleased) {
```
Comment (pleerock) [inline src/driver/cockroachdb/CockroachQueryRunner.ts:125]: what is the motivation behind this change? what was the problem caused these lines to be added?

File: src/query-builder/QueryBuilder.ts (hunk 1 above)
Comment (simplenotezy) [reply src/query-builder/QueryBuilder.ts:1046]: Good question; I implemented this while reading through their documentation that suggested casting it to string: https://www.cockroachlabs.com/docs/stable/array.html#supported-casting-and-conversion

**Comment by simplenotezy** [comment]
@pleerock I've also added support for time travelling queries. I am not 100% about whether the implementation is correct, but I have tested that it works at least.

**Comment by AlexMesser** [comment]
PR looks very good, but missing tests.

* for enums add `cockroachdb` in [here](https://github.com/typeorm/typeorm/blob/master/test/functional/database-schema/enums/enums.ts) and [here](https://github.com/typeorm/typeorm/blob/master/test/functional/database-schema/enums-array/enums-array.ts)(if enum arrays are supported by crdb)

File: docs/repository-api.md [hunk 3]
```diff
@@ -178,6 +178,7 @@ await repository.upsert(
     {
         conflictPaths: ["externalId"],
         skipUpdateIfNoValuesChanged: true, // supported by postgres, skips update if it would not change row values
+        upsertType: "upsert", //  "on-conflict-do-update" | "on-duplicate-key-update" | "upsert" - optionally provide an UpsertType - 'upsert' is currently only sypported by CockroachDB
```
Comment (AlexMesser) [inline docs/repository-api.md:181]: sypported -> supported

**Comment by dikshant** [comment]
@simplenotezy 👋🏽 I'm a product manager at Cockroach Labs overseeing our third party tooling support and stumbled on this. This is really great to see and we at Cockroach Labs really appreciate the effort you have put in here. Let me know if you need any help from us getting this to the finish line! Happy to help in any way we can.

**Comment by simplenotezy** [comment]
@dikshant thanks for the kind words! Happy to use Cockroachdb for our project, really exciting stuff.

So far it's been running in production for a while and it runs great. The PR can definitely be optimised a bit, as things was done quite quick while trying to get live on Cockroachdb. Anyhow, hope it'll get merged into TypeORM soon.

**Comment by dikshant** [comment]
@simplenotezy would you mind if someone from Cockroach Labs took a look at this PR and added the tests and the changes requested above?

**Comment by simplenotezy** [comment]
No @dikshant, that would be amazing. Exactly especially since we're pretty overwhelmed with work these days.
Let me know once he starts, and I'll hold any other updates I might otherwise bring

**Comment by simplenotezy** [comment]
The only thing I can recall right now I'm not that satisfied with it the implementation of:


->timeTravelQuery()

Example:

```
->timeTravelQuert(false) // disables it
->timeTravelQuery('-5m')
->timeTravelQuery(undefined) // leaves it on, this is the part I don't like, should rather be:
->timeTravelQuery(true) // uses follower_read_timestamp()
```

**Comment by dikshant** [comment]
@simplenotezy thanks! I am looking at getting some resources to work on this. 🔜

**Comment by AlexMesser** [comment]
* Added `enum` type support (original PR was missing code for synchronization) and tests
* Fixed issues with spatial types synchronization + tests
* Added support for spatial indices + tests
* Refactored Time Travel Queries functionality, removed TTQ support from FindOptions, added tests and docs.

File: test/functional/database-schema/column-types/cockroachdb/column-types-cockroach.ts [hunk 4]
```diff
@@ -28,6 +28,84 @@ describe("database schema > column types > cockroachdb", () => {
                 const table = await queryRunner.getTable("post")
                 await queryRunner.release()
 
+                const point = {
```
Comment (pleerock) [inline test/functional/database-schema/column-types/cockroachdb/column-types-cockroach.ts:31]: @AlexMesser we need to add additional docs on how to use these types.

File: src/query-builder/QueryBuilder.ts (hunk 1 above)
Comment (pleerock) [reply src/query-builder/QueryBuilder.ts:1067]: @AlexMesser what about this comment?

File: src/driver/DriverUtils.ts [hunk 5]
```diff
@@ -41,7 +41,9 @@ export class DriverUtils {
     }
 
     static isPostgresFamily(driver: Driver): boolean {
-        return ["postgres", "aurora-postgres"].includes(driver.options.type)
+        return ["postgres", "aurora-postgres", "cockroachdb"].includes(
```
Comment (pleerock) [inline src/driver/DriverUtils.ts:44]: @AlexMesser need to ensure all the previous `isPostgresFamily` calls work fine for the newly added driver.

File: test/functional/database-schema/column-types/cockroachdb/entity/Post.ts [hunk 6]
```diff
@@ -142,6 +142,56 @@ export class Post {
     @Column("inet")
     inet: string
 
+    // -------------------------------------------------------------------------
+    // Geometry Type
+    // -------------------------------------------------------------------------
+
+    @Column("geometry")
+    point: Object
```
Comment (pleerock) [inline test/functional/database-schema/column-types/cockroachdb/entity/Post.ts:150]: @AlexMesser what is the actual type for the "Object"?

File: test/functional/repository/basic-methods/repository-basic-methods.ts [hunk 7]
```diff
@@ -440,7 +440,7 @@ describe("repository > basic methods", () => {
         it("should first create then update an entity", () =>
             Promise.all(
                 connections.map(async (connection) => {
-                    if (connection.driver.supportedUpsertType == null) return
+                    if (!connection.driver.supportedUpsertTypes) return
```
Comment (pleerock) [inline test/functional/repository/basic-methods/repository-basic-methods.ts:443]: @AlexMesser best if we use empty arrays and remove undefined state from the `supportedUpsertTypes`. Less types - less problems.

File: test/functional/database-schema/column-types/cockroachdb/entity/Post.ts (hunk 6 above)
Comment (simplenotezy) [reply test/functional/database-schema/column-types/cockroachdb/entity/Post.ts:150]: GeoJSON/Point I believe

**Comment by AlexMesser** [comment]
* added [GeoJSON types](https://github.com/typeorm/typeorm/blob/564df2c294ba912d00637a6abce3147c9b49e6f3/src/driver/types/GeoJsonTypes.ts)
* updated [Spatial Columns](https://github.com/typeorm/typeorm/blob/b4e8daa0e63427677a60396842759e8fb5568bd9/docs/entities.md#spatial-columns) docs
* resolved review suggestions

**Comment by pleerock** [comment]
@simplenotezy you can test these changes in `0.3.12-dev.defb409`, e.g. by running `npm i typeorm@0.3.12-dev.defb409`

**Comment by simplenotezy** [comment]
@pleerock Apologies for late response. I'll check it out.

This might be a concern for a different PR/issue, but I noticed that our pods restart if the following error occours:

```
QueryFailedError: restart transaction: TransactionRetryWithProtoRefreshError: TransactionRetryError: retry txn (RETRY_SERIALIZABLE - failed preemptive refresh due to a conflict: committed value on key /Table/5832/2/"@"/"&\xd41\u007f/\xc3B\\\xbc\xaf\u008d\x9cm\xfe\xf5"/1916-12-06T11:43:05.437999999Z/"'̲\b4\xbaBM\x9d\xb0\x14\x1e\xceGM\x03"/0): "sql txn" meta={id=dab3e473 key=/Table/5609/2/"ple\x95@\x9aN\xf6\xbdJ\x99\xb0\xac0\xc3\x11"/"@"/"\v\xb5\xa5\x9f\r\x17@\x94\x95\xcd\x1c)#\x93\x1e$"/0 pri=0.02922444 epo=1 ts=1674735414.340096129,2 min=1674735412.057452290,0 seq=68} lock=true stat=PENDING rts=1674735412.192867432,1 wto=false gul=1674735412.307452290,0 at CockroachQueryRunner.query (/node_modules/typeorm/driver/cockroachdb/CockroachQueryRunner.js:225:19) at runMicrotasks (<anonymous>) at processTicksAndRejections (node:internal/process/task_queues:96:5) at async DeleteQueryBuilder.execute (/node_modules/typeorm/query-builder/DeleteQueryBuilder.js:52:33) at async ChatService.deleteChats (/dist/chat/service/chat.service.js:335:9)
```

The error is being triggered by this code:

```ts
    await this.repository
      .createQueryBuilder()
      .delete()
      .from(Chat)
      .where({
        id: In(ids),
      })
      .execute();
```

We are still using TypeORM `0.3.74` but we are planning on testing the latest fixed to cockroach ASAP.

**Comment by simplenotezy** [comment]
@AlexMesser is there a simple way to use `timeTravel`  in `FindOptions`? Or how would you recommend this be done?

E.g for queries like this:

```ts
await this.repository.count({ where: { chatId, userId }, timeTravel: true })

File: src/query-builder/SelectQueryBuilder.ts [hunk 8]
```diff
@@ -1326,24 +1326,13 @@ export class SelectQueryBuilder<Entity extends ObjectLiteral>
...
-        }
-
-        if (timestampFn !== false) {
-            this.expressionMap.timeTravel = true
-
-            if (timestampFn) {
-                this.expressionMap.timeTravelQueryTimestampFn = timestampFn
+    timeTravelQuery(timeTravelFn?: string | boolean): this {
+        if (this.connection.driver.options.type === "cockroachdb") {
+            if (timeTravelFn === undefined) {
```
Comment (simplenotezy) [inline src/query-builder/SelectQueryBuilder.ts:1331]: My only problem with this is that it create further complexity to the logic, if you wish to enable time travelling queries based on a condition.

E.g. it's not possible to do:

```
queryBuilder.timeTravelQuery(false) // disable all TTQ
queryBuilder.timeTravelQuery(shouldUseTTQ) // disable all TTQ based on condition
```

**Comment by pleerock** [comment]
@simplenotezy 

> This might be a concern for a different PR/issue, but I noticed that our pods restart if the following error occours:

We'll need a reproduction for this issue.

> is there a simple way to use timeTravel in FindOptions? Or how would you recommend this be done?

Using QueryBuilder. I do understand your concerns of using FindOptions instead of QueryBuilder, however right now we don't pollute FindOptions with driver-specific features. Time travel queries is a very specific feature.

**Comment by simplenotezy** [comment]
> @simplenotezy
> 
> > This might be a concern for a different PR/issue, but I noticed that our pods restart if the following error occours:
> 
> We'll need a reproduction for this issue.

I see. I'm not sure how to reproduce those errors, as they stem from CockroachDB and does not always occur, even when executing the exact same query several times in a row. Perhaps @AlexMesser or @dikshant can help elaborate.

> > is there a simple way to use timeTravel in FindOptions? Or how would you recommend this be done?
> 
> Using QueryBuilder. I do understand your concerns of using FindOptions instead of QueryBuilder, however right now we don't pollute FindOptions with driver-specific features. Time travel queries is a very specific feature.

I understand. I would be cool though if TS allowed setting extra FindOptions depending on the driver used.

Anyhow, it would be very helpful if there was a way conditionally to enable TTQ (TimeTravelQuery) without having to conditionally chain a method, but instead pass in a variable to either enable it or disable it, as I mentioned in my last post.

File: src/query-builder/SelectQueryBuilder.ts (hunk 8 above)
Comment (simplenotezy) [reply src/query-builder/SelectQueryBuilder.ts:1331]: Ah, well, I just briefly went through the code, and it seemed like it wasn't possible to do that, but if that's possible, well then that's great!

**Comment by simplenotezy** [comment]
> @pleerock Apologies for late response. I'll check it out.
> 
> This might be a concern for a different PR/issue, but I noticed that our pods restart if the following error occours:
> 
> ```
> QueryFailedError: restart transaction: TransactionRetryWithProtoRefreshError: TransactionRetryError: retry txn (RETRY_SERIALIZABLE - failed preemptive refresh due to a conflict: committed value on key /Table/5832/2/"@"/"&\xd41\u007f/\xc3B\\\xbc\xaf\u008d\x9cm\xfe\xf5"/1916-12-06T11:43:05.437999999Z/"'̲\b4\xbaBM\x9d\xb0\x14\x1e\xceGM\x03"/0): "sql txn" meta={id=dab3e473 key=/Table/5609/2/"ple\x95@\x9aN\xf6\xbdJ\x99\xb0\xac0\xc3\x11"/"@"/"\v\xb5\xa5\x9f\r\x17@\x94\x95\xcd\x1c)#\x93\x1e$"/0 pri=0.02922444 epo=1 ts=1674735414.340096129,2 min=1674735412.057452290,0 seq=68} lock=true stat=PENDING rts=1674735412.192867432,1 wto=false gul=1674735412.307452290,0 at CockroachQueryRunner.query (/node_modules/typeorm/driver/cockroachdb/CockroachQueryRunner.js:225:19) at runMicrotasks (<anonymous>) at processTicksAndRejections (node:internal/process/task_queues:96:5) at async DeleteQueryBuilder.execute (/node_modules/typeorm/query-builder/DeleteQueryBuilder.js:52:33) at async ChatService.deleteChats (/dist/chat/service/chat.service.js:335:9)
> ```
> 
> The error is being triggered by this code:
> 
> ```ts
>     await this.repository
>       .createQueryBuilder()
>       .delete()
>       .from(Chat)
>       .where({
>         id: In(ids),
>       })
>       .execute();
> ```
> 
> We are still using TypeORM `0.3.74` but we are planning on testing the latest fixed to cockroach ASAP.

Here's some info on that error: https://www.cockroachlabs.com/docs/stable/transaction-retry-error-reference.html#retry_serializable

It appears that they [recommend retrying the query](https://www.cockroachlabs.com/docs/v22.2/transactions#client-side-intervention).

Currently, when a `TransactionRetryWithProtoRefreshError` error is thrown, our app dies (Nest.js) causing a restart of the pod.

**Comment by simplenotezy** [comment]
> > @pleerock Apologies for late response. I'll check it out.
> > This might be a concern for a different PR/issue, but I noticed that our pods restart if the following error occours:
> > ```
> > QueryFailedError: restart transaction: TransactionRetryWithProtoRefreshError: TransactionRetryError: retry txn (RETRY_SERIALIZABLE - failed preemptive refresh due to a conflict: committed value on key /Table/5832/2/"@"/"&\xd41\u007f/\xc3B\\\xbc\xaf\u008d\x9cm\xfe\xf5"/1916-12-06T11:43:05.437999999Z/"'̲\b4\xbaBM\x9d\xb0\x14\x1e\xceGM\x03"/0): "sql txn" meta={id=dab3e473 key=/Table/5609/2/"ple\x95@\x9aN\xf6\xbdJ\x99\xb0\xac0\xc3\x11"/"@"/"\v\xb5\xa5\x9f\r\x17@\x94\x95\xcd\x1c)#\x93\x1e$"/0 pri=0.02922444 epo=1 ts=1674735414.340096129,2 min=1674735412.057452290,0 seq=68} lock=true stat=PENDING rts=1674735412.192867432,1 wto=false gul=1674735412.307452290,0 at CockroachQueryRunner.query (/node_modules/typeorm/driver/cockroachdb/CockroachQueryRunner.js:225:19) at runMicrotasks (<anonymous>) at processTicksAndRejections (node:internal/process/task_queues:96:5) at async DeleteQueryBuilder.execute (/node_modules/typeorm/query-builder/DeleteQueryBuilder.js:52:33) at async ChatService.deleteChats (/dist/chat/service/chat.service.js:335:9)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The error is being triggered by this code:
> > ```ts
> >     await this.repository
> >       .createQueryBuilder()
> >       .delete()
> >       .from(Chat)
> >       .where({
> >         id: In(ids),
> >       })
> >       .execute();
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > We are still using TypeORM `0.3.74` but we are planning on testing the latest fixed to cockroach ASAP.
> 
> Here's some info on that error: https://www.cockroachlabs.com/docs/stable/transaction-retry-error-reference.html#retry_serializable
> 
> It appears that they [recommend retrying the query](https://www.cockroachlabs.com/docs/v22.2/transactions#client-side-intervention).
> 
> Currently, when a `TransactionRetryWithProtoRefreshError` error is thrown, our app dies (Nest.js) causing a restart of the pod.
> 
> @AlexMesser / @dikshant - would you know how we could create a reproducible example to debug this further? I'm not entirely sure, I have been groping in the blind with something like this:
> 
> ```ts
>     await Promise.all(
>       Array(100)
>         .fill(0)
>         .map((_, i) =>
>           i % 2 === 0
>             ? this.dataSource.query('UPDATE "profile" SET last_active = NOW() WHERE user_id = $1', [
>                 'ec725f25-8821-4bb3-b263-0155e17e918f',
>               ])
>             : this.dataSource.query('SELECT last_active FROM "profile" WHERE user_id = $1', [
>                 'ec725f25-8821-4bb3-b263-0155e17e918f',
>               ]),
>         ),
>     );
> ```

I have created a dedicated issue for the retries here: https://github.com/typeorm/typeorm/issues/9984

**Comment by simplenotezy** [comment]
Shortly after creating the issue I discovered that a promise that would run the failing transaction wasn't awaited, and after awaiting it, it is now properly handled.

I'll keep the issue open since I am not sure if this is intended; i.e. queries that are not awaited would cause the application to crash, if they fail
